### PR TITLE
Bug/sync optimization logger across ranks

### PR DESCRIPTION
**Pull request by bmosaicml**
# What does this PR do?

image

Changing the optimization logger logic so that it correctly aggregates l2_norms, and cosines across ranks.

# What issue(s) does this change relate to?

Related to this bug.
https://mosaicml.atlassian.net/browse/CO-1787 

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Comment by mvpatel2000** [comment]
@bmosaicml please include a screenshot showing it works!

**Comment by eracah** [comment]
should we implement a `dist.reduce` function at some point, so we don't do all these all_reduces. Maybe doesn't make much of a difference given the message size, but....

**Comment by bmosaicml** [comment]
> should we implement a `dist.reduce` function at some point, so we don't do all these all_reduces? Maybe doesn't make much of a difference given the message size, but....

yeah I see very minimal throughput effects from the optimization monitor overall <1% so I suspect the effects of all_reduce vs reduce are infinitesimal?

File: tests/callbacks/test_optimizer_monitor.py [hunk 1]
```diff
@@ -34,15 +35,59 @@ def test_optimizer_monitor(log_optimizer_metrics: bool):
...
+        assert 'percentage_nonzero/second_moment/module.2.weight' in in_memory_logger.data.keys()
+
+    # expected to log gradient norm once per step (total batch)
+    assert grad_norm_calls == num_train_steps
+    for num_calls in layer_norm_calls:
+        assert num_calls == num_train_steps
+
+
+@device('gpu')
+@world_size(1, 4)
```
Comment (mvpatel2000) [inline tests/callbacks/test_optimizer_monitor.py:53]:
```suggestion
@world_size(1, 2)
```
We don't have 4 GPUs in CI/CD

File: composer/optim/decoupled_weight_decay.py [hunk 2]
```diff
@@ -350,6 +360,56 @@ def step(self, closure=None):
...
+                    dist.all_reduce(reduced, reduce_operation='SUM')
+
+                _, vectors, layer = tuple(metric.split('/'))
+
+                A, B = tuple(vectors.split('_'))
+
+                # it would've already been squared, so let's undo that
+                A_reduced_norm = optimizer_metrics[f'l2_norm/{A}/{layer}']
+                B_reduced_norm = optimizer_metrics[f'l2_norm/{B}/{layer}']
+                optimizer_metrics[metric] = reduced / (A_reduced_norm * B_reduced_norm)
```
Comment (mvpatel2000) [inline composer/optim/decoupled_weight_decay.py:383]: I don't understand the comment -- you're not square rooting anything here?

File: composer/optim/decoupled_weight_decay.py [hunk 3]
```diff
@@ -350,6 +360,56 @@ def step(self, closure=None):
...
+                reduced = optimizer_metrics[metric]
+                if dist.get_world_size() > 1:
+                    dist.all_reduce(reduced, reduce_operation='SUM')
+                optimizer_metrics[metric] = reduced / dist.get_world_size()
+
+        return optimizer_metrics
+
+    def pre_reduce_metrics(self, optimizer_metrics):
+        # some of the metrics need to be modified before being reduced in order for the
+        # reduction to work properly
```
Comment (mvpatel2000) [inline composer/optim/decoupled_weight_decay.py:394]:
```suggestion
        # Some metrics require preprocessing to reduce across ranks correctly
```

File: composer/optim/decoupled_weight_decay.py [hunk 4]
```diff
@@ -350,6 +360,56 @@ def step(self, closure=None):
...
+                # l2 norms need to be squared, before they are reduced via summation
+                optimizer_metrics[metric] = optimizer_metrics[metric]**2
+            elif metric.startswith('cosine'):
+                _, vectors, layer = tuple(metric.split('/'))
+
+                A, B = tuple(vectors.split('_'))
+
+                # it would've already been squared, so let's undo that
+                A_rank_subset_norm = math.sqrt(optimizer_metrics[f'l2_norm/{A}/{layer}'])
+                B_rank_subset_norm = math.sqrt(optimizer_metrics[f'l2_norm/{B}/{layer}'])
```
Comment (mvpatel2000) [inline composer/optim/decoupled_weight_decay.py:407]:
```suggestion
                # Take square root of squared metrics
                A_rank_subset_norm = math.sqrt(optimizer_metrics[f'l2_norm/{A}/{layer}'])
                B_rank_subset_norm = math.sqrt(optimizer_metrics[f'l2_norm/{B}/{layer}'])
```

**Review by mvpatel2000 — COMMENTED** [review]
nit: can we please start comments with a capital letter to be consistent with style?

Can unit test be updated to also include a run w/o FSDP?

File: tests/callbacks/test_optimizer_monitor.py [hunk 5]
```diff
@@ -34,15 +35,59 @@ def test_optimizer_monitor(log_optimizer_metrics: bool):
...
         assert 'l2_norm/moment/module.2.weight' in in_memory_logger.data.keys()
-        assert 'l2_norm_ratio/moment_grad/module.2.weight' in in_memory_logger.data.keys()
         assert 'cosine/moment_grad/module.2.weight' in in_memory_logger.data.keys()
         assert 'l2_norm/second_moment_sqrt/module.2.weight' in in_memory_logger.data.keys()
         assert 'l2_norm/update/module.2.weight' in in_memory_logger.data.keys()
         assert 'cosine/update_grad/module.2.weight' in in_memory_logger.data.keys()
-        assert 'l2_norm_ratio/update_param/module.2.weight' in in_memory_logger.data.keys()
+        assert 'percentage_nonzero/second_moment/module.2.weight' in in_memory_logger.data.keys()
+
+    # expected to log gradient norm once per step (total batch)
```
Comment (mvpatel2000) [inline tests/callbacks/test_optimizer_monitor.py:46]:
```suggestion
    # Expected to log gradient norm once per step (total batch)
```

File: tests/callbacks/test_optimizer_monitor.py [hunk 6]
```diff
@@ -34,15 +35,59 @@ def test_optimizer_monitor(log_optimizer_metrics: bool):
...
+    assert 'l2_norm/moment/module.2.weight' in in_memory_logger.data.keys()
+    assert 'l2_norm_ratio/moment_grad/module.2.weight' in in_memory_logger.data.keys()
+    assert 'cosine/moment_grad/module.2.weight' in in_memory_logger.data.keys()
+    assert 'l2_norm/second_moment_sqrt/module.2.weight' in in_memory_logger.data.keys()
+    assert 'l2_norm/update/module.2.weight' in in_memory_logger.data.keys()
+    assert 'cosine/update_grad/module.2.weight' in in_memory_logger.data.keys()
+    assert 'l2_norm_ratio/update_param/module.2.weight' in in_memory_logger.data.keys()
+    assert 'percentage_nonzero/second_moment/module.2.weight' in in_memory_logger.data.keys()
 
     # expected to log gradient norm once per step (total batch)
```
Comment (mvpatel2000) [inline tests/callbacks/test_optimizer_monitor.py:92]:
```suggestion
    # Expected to log gradient norm once per step (total batch)
```

**Review by mvpatel2000 — COMMENTED** [review]
more nits sorry for being annoying lol

File: composer/callbacks/optimizer_monitor.py [hunk 7]
```diff
@@ -87,21 +88,35 @@ def __init__(self, log_optimizer_metrics: bool = True):
...
 
-        default_metrics['l2_norm/grad/global'] = norm**0.5
+                if f'l2_norm/grad/{name}' not in optimizer_metrics:
+                    param_grad_norm = torch.linalg.vector_norm(p.grad)
+                    optimizer_metrics[f'l2_norm/grad/{name}'] = param_grad_norm
 
-        logger.log_metrics(default_metrics)
-        if self.log_optimizer_metrics:
-            logger.log_metrics(optimizer_metrics)
+        if state.fsdp_enabled and dist.get_world_size() > 0 and self.log_optimizer_metrics:
```
Comment (mvpatel2000) [inline composer/callbacks/optimizer_monitor.py:104]: Can you put a comment here explaining what's going in with FSDP?

File: composer/callbacks/optimizer_monitor.py [hunk 8]
```diff
@@ -87,21 +88,43 @@ def __init__(self, log_optimizer_metrics: bool = True):
...
                     optimizer_metrics = metric_reporter(p, name, optimizer_metrics)
 
-        default_metrics['l2_norm/grad/global'] = norm**0.5
-
-        logger.log_metrics(default_metrics)
-        if self.log_optimizer_metrics:
-            logger.log_metrics(optimizer_metrics)
+                if f'l2_norm/grad/{name}' not in optimizer_metrics:
+                    param_grad_norm = torch.linalg.vector_norm(p.grad)
+                    optimizer_metrics[f'l2_norm/grad/{name}'] = param_grad_norm
```
Comment (mvpatel2000) [inline composer/callbacks/optimizer_monitor.py:102]:
```suggestion
                # Always log grad norm as a default metric if it's not specified
                if f'l2_norm/grad/{name}' not in optimizer_metrics:
                    param_grad_norm = torch.linalg.vector_norm(p.grad)
                    optimizer_metrics[f'l2_norm/grad/{name}'] = param_grad_norm
```

File: composer/optim/decoupled_weight_decay.py [hunk 9]
```diff
@@ -350,6 +349,54 @@ def step(self, closure=None):
...
+            else:
+                reduced = optimizer_metrics[metric]
+                if dist.get_world_size() > 1:
+                    dist.all_reduce(reduced, reduce_operation='SUM')
+                optimizer_metrics[metric] = reduced / dist.get_world_size()
+
+        return optimizer_metrics
+
+    def pre_reduce_metrics(self, optimizer_metrics):
+        # Some metrics require preprocessing to reduce across ranks correctly
```
Comment (mvpatel2000) [inline composer/optim/decoupled_weight_decay.py:381]:
```suggestion
    def pre_reduce_metrics(self, optimizer_metrics):
        '''Preprocess metrics to reduce across ranks correctly.'''
```

File: composer/optim/decoupled_weight_decay.py [hunk 10]
```diff
@@ -350,6 +349,54 @@ def step(self, closure=None):
...
+            elif metric.startswith('cosine'):
+                _, vectors, layer = tuple(metric.split('/'))
+
+                A, B = tuple(vectors.split('_'))
+
+                # L2 norm would've been squared in previous branch
+                A_rank_subset_norm = math.sqrt(optimizer_metrics[f'l2_norm/{A}/{layer}'])
+                B_rank_subset_norm = math.sqrt(optimizer_metrics[f'l2_norm/{B}/{layer}'])
+
+                optimizer_metrics[metric] *= A_rank_subset_norm * B_rank_subset_norm
```
Comment (mvpatel2000) [inline composer/optim/decoupled_weight_decay.py:396]: I don't think this works -- you assume L2 norms are processed and squared before cosine norms

Comment (bmosaicml) [reply composer/optim/decoupled_weight_decay.py:396]: They should be iterated in order of the key names in the class' metric_functions!

Comment (mvpatel2000) [reply composer/optim/decoupled_weight_decay.py:395]: Hm... that's a little hacky... I would prefer to explicitly sort it. Can we do something like:

```
# Sort L2 norms first so they are squared before other metrics, which depend on squared values
metrics = optimizer_metrics.keys()
metrics = sorted(metrics, key=lambda metric: 0 if 'l2_norm' in metric else 1)  
```

and then put a comment here that says "L2 norms have been already processed and squared"

File: composer/optim/decoupled_weight_decay.py [hunk 11]
```diff
@@ -350,6 +350,56 @@ def step(self, closure=None):
...
+        for metric in metrics:
+            if metric.startswith('l2_norm'):
+                # L2 norms need to be squared, before they are reduced via summation
+                optimizer_metrics[metric] = optimizer_metrics[metric]**2
+            elif metric.startswith('cosine'):
+                _, vectors, layer = tuple(metric.split('/'))
+
+                A, B = tuple(vectors.split('_'))
+
+                # L2 norm would've been squared in previous branch
```
Comment (mvpatel2000) [inline composer/optim/decoupled_weight_decay.py:395]:
```suggestion
                # L2 norms are already squared as they are sorted first in list
```

File: composer/optim/decoupled_weight_decay.py [hunk 12]
```diff
@@ -350,6 +350,56 @@ def step(self, closure=None):
...
+        return optimizer_metrics
+
+    def pre_reduce_metrics(self, optimizer_metrics):
+        """Preprocess metrics to reduce across ranks correctly."""
+        # Sort L2 norms first so they are squared before other metrics, which depend on squared values
+        metrics = optimizer_metrics.keys()
+        metrics = sorted(metrics, key=lambda metric: 0 if 'l2_norm' in metric else 1)
+        for metric in metrics:
+            if metric.startswith('l2_norm'):
+                # L2 norms need to be squared, before they are reduced via summation
```
Comment (mvpatel2000) [inline composer/optim/decoupled_weight_decay.py:388]:
```suggestion
                # L2 norms need to be squared before they are reduced via summation
```

**Review by mvpatel2000 — APPROVED** [review]

**Pull request merged**